homeassistant/components/heos/media_player.py

```diff
@@ -0,0 +1,203 @@
+    "pause": STATE_PAUSED,
+    "stop": STATE_IDLE,
+}
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discover_info=None):
+    """Set up the HEOS platform."""
+    controller = hass.data[HEOS_DOMAIN][DOMAIN]
+    players = controller.get_players()
+    devices = [HeosMediaPlayer(p) for p in players]
+    async_add_devices(devices, True)
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, player):
+        """Initialize."""
+        self._player = player
+        self._position_jitter_acceptance_ms = 500
+        self._cache_position_ms = 0
+        self._cache_position_at = utcnow()
+
+    def _update_state(self):
+        self.async_schedule_update_ha_state()
+
+    async def async_update(self):
+        """Update the player."""
+        self._player.request_update()
+
+    async def async_added_to_hass(self):
+        """Device added to hass."""
+        self._player.state_change_callback = self._update_state
+
+    @property
+    def unique_id(self):
+        """Get unique id of the player."""
+        return self._player.player_id
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._player.name
+
+    @property
+    def volume_level(self):
+        """Volume level of the device (0..1)."""
+        volume = self._player.volume
+        return float(volume) / 100
+
+    @property
+    def state(self):
+        """Get state."""
+        return PLAY_STATE_TO_STATE.get(self._player.play_state)
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def media_content_type(self):
+        """Content type of current playing media."""
+        return MEDIA_TYPE_MUSIC
+
+    @property
+    def media_artist(self):
+        """Artist of current playing media."""
+        return self._player.media_artist
+
+    @property
+    def media_title(self):
+        """Album name of current playing media."""
+        return self._player.media_title
+
+    @property
+    def media_album_name(self):
+        """Album name of current playing media."""
+        return self._player.media_album
+
+    @property
+    def media_image_url(self):
+        """Return the image url of current playing media."""
+        return self._player.media_image_url
+
+    @property
+    def media_content_id(self):
+        """Return the content ID of current playing media."""
+        return self._player.media_id
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self._player.mute == "on"
+
+    async def async_mute_volume(self, mute):
+        """Mute volume."""
+        self._player.set_mute(mute)
+
+    @property
+    def media_duration(self):
+        """Duration of current playing media in seconds."""
+        return self._player.duration / 1000.0
+
+    def _get_cache_position(self):
```

Comment:
While this freezes the values of position, this isn't stopping the callback from updating the state which is where the issue lies.  As I understand the requirements (@awarecan please confirm):
- Set `media_position_updated_at` and `media_position` only upon transition _to_ `STATE_PLAYING` or when the media changes (there's no seek support in HEOS).
- The front-end automatically advances the position based on the state (playing, paused, stopped).
- Do not update state or position properties upon subsequent media position change (which HEOS sends every second)

The library is the best place to implement this logic since it has access to the lower level events needed to determine how to issue updates.

Reply:
First, I was kind of careless when I was implemented the position update. I thought this (https://github.com/home-assistant/architecture/issues/101) was implemented. And it was because of this I did it in this way.

Second, I still believe that this issue should be implemented in HA because the problem lies in HA. The library should be front end agnostic.